File: packages/vm/wasmlib/ts/wasmlib/keys.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// conversion of data bytes to key id

import {getKeyIDFromString} from "./host";

export interface MapKey {
    getKeyID(): Key32;
}

export class Key32 implements MapKey {
    keyID: i32;

    constructor(keyID: i32) {
        this.keyID = keyID;
    }

    static fromString(key: string): Key32 {
        return getKeyIDFromString(key);
    }

    getKeyID(): Key32 {
        return this;
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// all predefined key id values should exactly match their counterpart values on the host!
// note that predefined key ids are negative values to distinguish them from indexes

// @formatter:off
export const KEY_ACCOUNT_ID       = new Key32(-1);
export const KEY_ADDRESS          = new Key32(-2);
export const KEY_BALANCES         = new Key32(-3);
export const KEY_BASE58_DECODE    = new Key32(-4);
export const KEY_BASE58_ENCODE    = new Key32(-5);
export const KEY_BLS_ADDRESS      = new Key32(-6);
export const KEY_BLS_AGGREGATE    = new Key32(-7);
export const KEY_BLS_VALID        = new Key32(-8);
export const KEY_CALL             = new Key32(-9);
export const KEY_CALLER           = new Key32(-10);
export const KEY_CHAIN_ID         = new Key32(-11);
export const KEY_CHAIN_OWNER_ID   = new Key32(-12);
export const KEY_COLOR            = new Key32(-13);
export const KEY_CONTRACT         = new Key32(-14);
export const KEY_CONTRACT_CREATOR = new Key32(-15);
export const KEY_DEPLOY           = new Key32(-16);
export const KEY_ED25519_ADDRESS  = new Key32(-17);
export const KEY_ED25519_VALID    = new Key32(-18);
export const KEY_EVENT            = new Key32(-19);
export const KEY_EXPORTS          = new Key32(-20);
export const KEY_HASH_BLAKE2B     = new Key32(-21);
export const KEY_HASH_SHA3        = new Key32(-22);
export const KEY_HNAME            = new Key32(-23);
export const KEY_INCOMING         = new Key32(-24);
export const KEY_LENGTH           = new Key32(-25);
export const KEY_LOG              = new Key32(-26);
export const KEY_MAPS             = new Key32(-27);
export const KEY_MINTED           = new Key32(-28);
export const KEY_PANIC            = new Key32(-29);
export const KEY_PARAMS           = new Key32(-30);
export const KEY_POST             = new Key32(-31);
export const KEY_RANDOM           = new Key32(-32);
export const KEY_REQUEST_ID       = new Key32(-33);
export const KEY_RESULTS          = new Key32(-34);
export const KEY_RETURN           = new Key32(-35);
export const KEY_STATE            = new Key32(-36);
export const KEY_TIMESTAMP        = new Key32(-37);
export const KEY_TRACE            = new Key32(-38);
export const KEY_TRANSFERS        = new Key32(-39);
export const KEY_UTILITY          = new Key32(-40);
export const KEY_ZZZZZZZ          = new Key32(-41);
// @formatter:on

File: packages/vm/wasmlib/go/wasmlib/keys.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

type MapKey interface {
	KeyID() Key32
}

type Key string

func (key Key) KeyID() Key32 {
	return GetKeyIDFromString(string(key))
}

type Key32 int32

func (key Key32) KeyID() Key32 {
	return key
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const (
	KeyAccountID       = Key32(-1)
	KeyAddress         = Key32(-2)
	KeyBalances        = Key32(-3)
	KeyBase58Decode    = Key32(-4)
	KeyBase58Encode    = Key32(-5)
	KeyBlsAddress      = Key32(-6)
	KeyBlsAggregate    = Key32(-7)
	KeyBlsValid        = Key32(-8)
	KeyCall            = Key32(-9)
	KeyCaller          = Key32(-10)
	KeyChainID         = Key32(-11)
	KeyChainOwnerID    = Key32(-12)
	KeyColor           = Key32(-13)
	KeyContract        = Key32(-14)
	KeyContractCreator = Key32(-15)
	KeyDeploy          = Key32(-16)
	KeyEd25519Address  = Key32(-17)
	KeyEd25519Valid    = Key32(-18)
	KeyEvent           = Key32(-19)
	KeyExports         = Key32(-20)
	KeyHashBlake2b     = Key32(-21)
	KeyHashSha3        = Key32(-22)
	KeyHname           = Key32(-23)
	KeyIncoming        = Key32(-24)
	KeyLength          = Key32(-25)
	KeyLog             = Key32(-26)
	KeyMaps            = Key32(-27)
	KeyMinted          = Key32(-28)
	KeyPanic           = Key32(-29)
	KeyParams          = Key32(-30)
	KeyPost            = Key32(-31)
	KeyRandom          = Key32(-32)
	KeyRequestID       = Key32(-33)
	KeyResults         = Key32(-34)
	KeyReturn          = Key32(-35)
	KeyState           = Key32(-36)
	KeyTimestamp       = Key32(-37)
	KeyTrace           = Key32(-38)
	KeyTransfers       = Key32(-39)
	KeyUtility         = Key32(-40)
	KeyZzzzzzz         = Key32(-41)
)
